contracts/libraries/LibArray.sol

```diff
@@ -0,0 +1,315 @@
+    uint256 length = ids.length;
+    result = new uint256[](length);
+    for (uint256 i; i < length; ++i) {
+      result[i] = self[ids[i]];
+    }
+  }
+
+  /**
+   * @dev Picks the top `k` ids based on their corresponding values.
+   * @param ids The array of ids.
+   * @param values The array of values corresponding to the ids.
+   * @param k The number of ids to pick.
+   * @return pickeds The array of the top `k` picked ids.
+   * @return numNotPicked The number of ids that were not picked.
+   */
+  function pickTopKByValues(
+    uint256[] memory ids,
+    uint256[] memory values,
+    uint256 k
+  ) internal pure returns (uint256[] memory pickeds, uint256 numNotPicked) {
+    unchecked {
+      uint256 length = ids.length;
+      if (k >= length) return (ids, 0);
+
+      inplaceSortByValue(ids, values);
+      inplaceResize(ids, k);
+
+      return (ids, length - k);
+    }
+  }
+
+  /**
+   * @dev Filter an array of address given array of indices.
+   * @param self The array of address.
+   * @param indices The array of indices to filter.
+   * @return result The resulting array of address.
+   */
+  function take(address[] memory self, uint256[] memory indices) internal pure returns (address[] memory result) {
+    return toAddresses(take(toUint256s(self), indices));
+  }
+
+  /**
+   * @dev Filters the given array of keys based on the provided filter function.
+   * @param keys The array of keys to filter.
+   * @param v The array of values to pass to the filter function.
+   * @param filterFn The filter function that takes a key and an array of values and returns a boolean.
+   * @return filteredIds The array of filtered keys.
+   * @return nFiltered The number of filtered keys.
+   */
+  function filter(
+    uint256[] memory keys,
+    uint256[] memory v,
+    function(uint256, uint256[] memory) pure returns (bool) filterFn
+  ) internal pure returns (uint256[] memory filteredIds, uint256 nFiltered) {
+    unchecked {
+      uint256 length = keys.length;
+      filteredIds = new uint256[](length);
+
+      for (uint256 i; i < length; ++i) {
+        if (filterFn(i, v)) {
+          filteredIds[nFiltered] = keys[i];
+          ++nFiltered;
+        }
+      }
+
+      inplaceResize(filteredIds, nFiltered);
+
+      return (filteredIds, nFiltered);
+    }
+  }
+
+  function inplaceSortByValue(
+    address[] memory self,
+    uint256[] memory values
+  ) internal pure returns (address[] memory sorted) {
+    return toAddresses(inplaceQuickSortByValue(toUint256s(self), values));
+  }
+
+  /**
+   * @dev inplace resizes the given array to the specified length.
+   * @param self The array to be resized.
+   * @param length The new length of the array.
+   * @return resized The resized array.
+   * @notice This function modify `self`
+   */
+  function inplaceResize(uint256[] memory self, uint256 length) internal pure returns (uint256[] memory resized) {
```

Comment:
```suggestion
   * @dev Resize a memory array.
   *
   * WARNING: The new length of the array should not be greater than the current length to avoid collision with other already allocated memory.
   */
  function unsafeResize(uint256[] memory self, uint256 length) internal pure returns (uint256[] memory resized) {
```